### ISSUE TITLE
TSQL is being misclassified as PLpgSQL

### ISSUE THREAD

**Issue by beau-witter** [github/linguist#4476]
## Preliminary Steps

Please confirm you have...
- [X] reviewed [How Linguist Works](https://github.com/github/linguist#how-linguist-works),
- [X] reviewed the [Troubleshooting](https://github.com/github/linguist#troubleshooting) docs,
- [X] considered implementing an [override](https://github.com/github/linguist#overrides),
- [X] verified an issue has not already been logged for your issue ([linguist issues](https://github.com/issues?utf8=%E2%9C%93&q=is%3Aissue+repo%3Agithub/linguist)).

## Problem Description

Any Microsoft SQL (T-SQL) file is classified as a PLpgSQL file.

### URL of the affected repository:
Affected repository is private, so I cannot link here, but I found another repository that highlights this issue. https://github.com/SQL-Server-projects/TSQL-Scripts

That repository is purely for Microsoft SQL-Server TSQL code, but it is all classified as PLpgSQL.

### Last modified on:

Private repository is modified regularly, but the linked repository was updated 2019-03-28.

### Expected language:

Expected: T-SQL

### Detected language:

Detected: PLpgSQL

**Comment by lildude** [comment]
That'll be because Linguist doesn't know about the T-SQL language so is using the `.sql` extension to try an associate the files with one of the SQL languages it does know about and is determining they look more like PLpgSQL than any of the other options.

If you would like T-SQL recognised, please review the [contributing guidelines](https://github.com/github/linguist/blob/master/CONTRIBUTING.md) and feel free to submit a pull request.

Note, as this is another language associated with the `.sql` extension, you will need to extend the heuristics at https://github.com/github/linguist/blob/3feaaabfa16449d53ba3933f8bfba021d093bc6d/lib/linguist/heuristics.yml#L382-L395 to uniquely identify these files as T-SQL.

**Comment by beau-witter** [comment]
@lildude I think I have updated the heuristics and added samples correctly, but I am having trouble checking the output using `bundle exec bin/github-linguist --breakdown` as specified in the .README file.

I get `No such file or directory @ rb_sysopen - /mnt/c/Home/Linguist/lib/linguist/samples.json (Errno::ENOENT)`

I am running it through bash on Win10. Should I open a PR with my changes to have someone look over them? Or is there a way I can run this locally?

**Comment by lildude** [comment]
@beau-witter sounds like you missed the `bundle exec rake samples` step.

**Comment by beau-witter** [comment]
@lildude Sorry, I am such a noob, I went back and reran the `script/bootstrap` step at the very beginning. It shows no output (I am assuming it is running correctly) but then when I run `bundle exec rake samples` I get `rake aborted!
Errno::EISDIR: Is a directory @ io_fread - /mnt/c/Home/Linguist/samples/XML/bin`

I tried to look up what causes that, but I am very unfamiliar with Ruby.

Thanks in advance.

**Comment by lildude** [comment]
🤔 That error suggests there's a `bin` directory in your `/mnt/c/Home/Linguist/samples/XML` directory. There shouldn't be:

```
$ /bin/ls -l samples/XML/bin
ls: samples/XML/bin: No such file or directory
$
```

Delete `/mnt/c/Home/Linguist/samples/XML/bin` and assuming there are no other unexpected directories, the samples should generate as expected.

You might want to run `git status` and make sure there aren't any other unexpected files and directories in your `/mnt/c/Home/Linguist` directory.

**Comment by beau-witter** [comment]
@lildude Thanks so much for your help! I got all the way to Step 2 of Adding a Language (I had incorrectly been following the steps for Fixing a misclassified language) but I have hit another snag. I am running `script/add-grammar https://github.com/antlr/grammars-v4/tree/master/tsql` with that being my chosen grammar (I think it looked correct, I was unsure if the repo needed to be in a specific format). I always get an error when it gets to running the `script/grammar-compiler` file.

root@RL028984471853:/mnt/c/Home/Linguist# script/add-grammar https://github.com/antlr/grammars-v4/tree/master/tsql
Checking docker is installed and running
$ docker ps
Registering new submodule: vendor/grammars/grammars-v4
$ git submodule add -f https://github.com/antlr/grammars-v4 vendor/grammars/grammars-v4
$ script/grammar-compiler add vendor/grammars/grammars-v4
  > latest: Pulling from linguist/grammar-compiler
  > Digest: sha256:2c49232d2c7c06974c2908fe613630c3bb876f0742a12520440a1a4f98aa2e42
  > Status: Image is up to date for linguist/grammar-compiler:latest
  > open /src/linguist/lib/linguist/VERSION: no such file or directory
Command failed. Aborting.

I then used `docker export $(docker ps -lq) | tar tf - | grep VERSION` to see if the docker container even had the file it was looking for. Here is the result:

root@RL028984471853:/mnt/c/Home/Linguist# docker export $(docker ps -lq) | tar tf - | grep VERSION
usr/local/go/VERSION
usr/share/cmake-3.7/Help/prop_gbl/FIND_LIBRARY_USE_OPENBSD_VERSIONING.rst
usr/share/cmake-3.7/Help/prop_tgt/FRAMEWORK_VERSION.rst
usr/share/cmake-3.7/Help/prop_tgt/SOVERSION.rst
usr/share/cmake-3.7/Help/prop_tgt/VERSION.rst
usr/share/cmake-3.7/Help/prop_tgt/VS_DESKTOP_EXTENSIONS_VERSION.rst
usr/share/cmake-3.7/Help/prop_tgt/VS_DOTNET_TARGET_FRAMEWORK_VERSION.rst
usr/share/cmake-3.7/Help/prop_tgt/VS_IOT_EXTENSIONS_VERSION.rst
usr/share/cmake-3.7/Help/prop_tgt/VS_MOBILE_EXTENSIONS_VERSION.rst
usr/share/cmake-3.7/Help/prop_tgt/VS_WINDOWS_TARGET_PLATFORM_MIN_VERSION.rst
usr/share/cmake-3.7/Help/variable/CMAKE_ANDROID_NDK_TOOLCHAIN_VERSION.rst
usr/share/cmake-3.7/Help/variable/CMAKE_CACHE_MAJOR_VERSION.rst
usr/share/cmake-3.7/Help/variable/CMAKE_CACHE_MINOR_VERSION.rst
usr/share/cmake-3.7/Help/variable/CMAKE_CACHE_PATCH_VERSION.rst
usr/share/cmake-3.7/Help/variable/CMAKE_ECLIPSE_VERSION.rst
usr/share/cmake-3.7/Help/variable/CMAKE_HOST_SYSTEM_VERSION.rst
usr/share/cmake-3.7/Help/variable/CMAKE_LANG_COMPILER_VERSION.rst
usr/share/cmake-3.7/Help/variable/CMAKE_LANG_SIMULATE_VERSION.rst
usr/share/cmake-3.7/Help/variable/CMAKE_MAJOR_VERSION.rst
usr/share/cmake-3.7/Help/variable/CMAKE_MINIMUM_REQUIRED_VERSION.rst
usr/share/cmake-3.7/Help/variable/CMAKE_MINOR_VERSION.rst
usr/share/cmake-3.7/Help/variable/CMAKE_PATCH_VERSION.rst
usr/share/cmake-3.7/Help/variable/CMAKE_SYSTEM_VERSION.rst
usr/share/cmake-3.7/Help/variable/CMAKE_Swift_LANGUAGE_VERSION.rst
usr/share/cmake-3.7/Help/variable/CMAKE_TWEAK_VERSION.rst
usr/share/cmake-3.7/Help/variable/CMAKE_VERSION.rst
usr/share/cmake-3.7/Help/variable/CMAKE_VS_INTEL_Fortran_PROJECT_VERSION.rst
usr/share/cmake-3.7/Help/variable/CMAKE_VS_NsightTegra_VERSION.rst
usr/share/cmake-3.7/Help/variable/CMAKE_VS_WINDOWS_TARGET_PLATFORM_VERSION.rst
usr/share/cmake-3.7/Help/variable/CTEST_UPDATE_VERSION_ONLY.rst
usr/share/cmake-3.7/Help/variable/MSVC_VERSION.rst
usr/share/cmake-3.7/Help/variable/PROJECT-NAME_VERSION.rst
usr/share/cmake-3.7/Help/variable/PROJECT-NAME_VERSION_MAJOR.rst
usr/share/cmake-3.7/Help/variable/PROJECT-NAME_VERSION_MINOR.rst
usr/share/cmake-3.7/Help/variable/PROJECT-NAME_VERSION_PATCH.rst
usr/share/cmake-3.7/Help/variable/PROJECT-NAME_VERSION_TWEAK.rst
usr/share/cmake-3.7/Help/variable/PROJECT_VERSION.rst
usr/share/cmake-3.7/Help/variable/PROJECT_VERSION_MAJOR.rst
usr/share/cmake-3.7/Help/variable/PROJECT_VERSION_MINOR.rst
usr/share/cmake-3.7/Help/variable/PROJECT_VERSION_PATCH.rst
usr/share/cmake-3.7/Help/variable/PROJECT_VERSION_TWEAK.rst
usr/share/cmake-3.7/Help/variable/XCODE_VERSION.rst

I have already followed all the previous steps and ensured all the dependencies are up to date:

root@RL028984471853:/mnt/c/Home/Linguist# apt-get install cmake pkg-config libicu-dev docker-ce
Reading package lists... Done
Building dependency tree
Reading state information... Done
pkg-config is already the newest version (0.29.1-0ubuntu1).
cmake is already the newest version (3.5.1-1ubuntu3).
libicu-dev is already the newest version (55.1-7ubuntu0.4).
docker-ce is already the newest version (5:18.09.4~3-0~ubuntu-xenial).
0 upgraded, 0 newly installed, 0 to remove and 204 not upgraded.

What else can I do to get that step to pass?

**Comment by pchaigno** [comment]
Could you try `script/add-grammar https://github.com/antlr/grammars-v4` instead?

**Comment by beau-witter** [comment]
@pchaigno Hey, thanks for the suggestion! I just gave that a try and got the same thing as trying https://github.com/antlr/grammars-v4/tree/master/tsql. I could try a different grammar repo, but the problem seems to be more to do with the image, no? Could it be a type mismatch? I am using version 1.12.1 of Go, but it looks like the image is using 1.9.2.

**Comment by lildude** [comment]
> Could it be a type mismatch? I am using version 1.12.1 of Go, but it looks like the image is using 1.9.2.

That shouldn't matter as the compiler runs in docker and uses the version defined in the Dockerfile.

The issue here is your docker image doesn't appear to be finding the `lib/linguist/VERSION` file, ie this file - https://github.com/github/linguist/blob/master/lib/linguist/VERSION suggesting this line https://github.com/github/linguist/blob/c8ac505c26fa424cd3322eae8b455e8fb9b643cb/script/grammar-compiler#L19 ... is not taking effect, ie your repo isn't being mounted in the docker container.

I've never used Docker on Windows so I'm not sure how to debug this in your env. Sorry.

That all said, you're not going to be able to use https://github.com/antlr/grammars-v4/tree/master/tsql or even the repo as a whole as it doesn't appear to contain any TextMate-compatible grammars, so even once you resolve the docker issue, the compiler won't find anything...

```
$ script/add-grammar https://github.com/antlr/grammars-v4
Checking docker is installed and running
$ docker ps
Registering new submodule: vendor/grammars/grammars-v4
$ git submodule add -f https://github.com/antlr/grammars-v4 vendor/grammars/grammars-v4
$ script/grammar-compiler add vendor/grammars/grammars-v4
  > latest: Pulling from linguist/grammar-compiler
  > Digest: sha256:2c49232d2c7c06974c2908fe613630c3bb876f0742a12520440a1a4f98aa2e42
  > Status: Image is up to date for linguist/grammar-compiler:latest
  > source 'vendor/grammars/grammars-v4' contains no grammar files
$
```

Something like https://github.com/Microsoft/vscode-mssql for VS Code or https://github.com/mattiascibien/language-tsql for Atom would certainly work though.

**Comment by beau-witter** [comment]
@lildude I will try again from the beginning in a pure Linux environment (I will have to do some setup on another computer to get there though).

Thanks for the advice on the grammar repo selection though, I will try one of the other ones you suggested (I wasn't sure how to find them or what I was looking for).

Thank you so much! You both have been very helpful!

**Comment by beau-witter** [comment]
@lildude I progressed much further with the linux system attempt.

The problem I am getting now is during test running. I have 3 failing tests that I am unsure how to fix.

  1) Failure:
TestHeuristics#test_sql_by_heuristics [/home/USER/Code/linguist/test/test_heuristics.rb:434]:
Failed for SQL/create_stuff.sql.
--- expected
+++ actual
@@ -1 +1 @@
-[#<Linguist::Language name=SQL>]
+[#<Linguist::Language name=TSQL>]


  2) Failure:
TestFileBlob#test_language [/home/USER/Code/linguist/test/test_file_blob.rb:655]:
create_stuff.sql.
--- expected
+++ actual
@@ -1 +1,2 @@
-"SQL"
+# encoding: UTF-8
+"TSQL"


  3) Failure:
TestBlob#test_language [/home/USER/Code/linguist/test/test_blob.rb:263]:
create_stuff.sql.
--- expected
+++ actual
@@ -1 +1,2 @@
-"SQL"
+# encoding: UTF-8
+"TSQL"

I do not know why it is being classified like that though, in the heuristics I have negative patterns that should exclude some of the keywords in the create_stuff.sql. Could you help with me that?

Would you be able to help me with that?

**Comment by lildude** [comment]
These failures suggest your changes are now causing the https://github.com/github/linguist/blob/master/samples/SQL/create_stuff.sql file to be classified as TSQL. Either this is because your heuristics are wrong (ie associating generic SQL terms exclusively with TSQL) or the sample is actually TSQL so the tests need changing. Looking at the file itself, it looks like standard SQL to me thus suggesting your heuristics are misclassifying normal SQL files as TSQL.

It's worth pointing out that Linguist works on files in isolation with no attention paid to the repo or other files. So the heuristics need to be specific for that language which means if standard SQL is used for TSQL the file should be classified as SQL and not TSQL. TSQL should only be used to identify those files that are specifically TSQL and not another form of SQL. If a users really wants a standard SQL to be identified as TSQL, they should use an override… which of course depends on Linguist knowing about TSQL (which you're aiming to add).

I think the best thing to do is open a [draft PR](https://github.blog/2019-02-14-introducing-draft-pull-requests/) and we can comment directly on it at the appropriate lines. I've already taken a peek at your local changes and have a few comments and questions which are easier to do in context, which I can do in a PR.

**Comment by beau-witter** [comment]
@lildude [Here is the draft PR](https://github.com/search?utf8=%E2%9C%93&type=Code&ref=searchresults&q=extension%3Asql+PROCEDURE). I went through the heuristics I added again and compared them to the create_stuff.sql file, and I personally don't see how it could match, especially with the negative_pattern I added. However, I am not very familiar with the ruby regex, so I cannot say for certain I wrote it at all correctly. I look forward to seeing how I can make changes to get this working, you all have been a huge help!
